### DISCUSSION TITLE
Updating texture using scene graph does not apply for usdz models in AR

### DISCUSSION THREAD

**Discussion by pzyu** [google/model-viewer#2248] [Ideas]
### Description

I'm trying to update a texture on runtime based on this example: https://modelviewer.dev/examples/scenegraph/#swapTextures

This works perfectly in the browser and in AR mode for GLB models, however, the texture is not reflected for USDZ models in QuickLook.

Expected texture applied on the model:
![image](https://user-images.githubusercontent.com/18459487/112798153-161aff80-909f-11eb-9ccc-5a7650961030.png)

Resulting texture shown in iOS Quick Look:
![image](https://user-images.githubusercontent.com/18459487/112798113-056a8980-909f-11eb-8e8c-696166cd1998.png)

Is this something that's out of model-viewer's control? I understand that scene graph is only supported for GLTF files, but I'm not sure if it is supposed to work the same for USDZ models.

### Version

- model-viewer: v1.6.0

### Browser Affected

- [ ] Chrome, version: xx.x.xxxx.xx
- [ ] Edge
- [ ] Firefox
- [ ] IE
- [x] Safari

### OS

- [ ] Android
- [x] iOS
- [ ] Linux
- [ ] MacOS
- [ ] Windows

### AR

- [ ] WebXR
- [ ] SceneViewer
- [x] QuickLook

**Comment by elalish** [comment]
No, we don't make any changes to the USDZ file. However, we're working on doing on-the-fly conversion to USDZ which would mean both these kinds of changes would be reflected and also that you wouldn't need to supply a separate file. No ETA yet though.

**Comment by pzyu** [comment]
Sure, I see. Thanks for the reply. Since model-viewer is built on three.js, I'm thinking of using the USDZ exporter that they have as shown in the example here: https://threejs.org/examples/?q=usdz#misc_exporter_usdz

Would it be possible to export the model as a GLTF and then export it as USDZ using three.js?

**Comment by pzyu** [comment]
> 
> 
> Sure, I see. Thanks for the reply. Since model-viewer is built on three.js, I'm thinking of using the USDZ exporter that they have as shown in the example here: https://threejs.org/examples/?q=usdz#misc_exporter_usdz
> 
> Would it be possible to export the model as a GLTF and then export it as USDZ using three.js?

So I did manage to get it to work using three.js' USDZ exporter by doing the following:
 1. Export the model as GLTF
 2. Load it into another three.js scene as an array buffer
 3. Export the scene using the USDZ exporter as a blob
 4. Setting the href property of a link as the blob

Working result!
-----
Currently the texture is flipped, but shouldn't be much of a problem to resolve.

![image](https://user-images.githubusercontent.com/18459487/112956634-416f1e80-9173-11eb-81a1-8517c291908f.png)

**Comment by elalish** [comment]
Yes, that's exactly how we'd like to enable on-the-fly conversion, nice work! Would you be interested in contributing?

**Comment by pzyu** [comment]
Sure, will make a fork and send a PR accordingly. 

One thing to note is that for my particular use case, I've edited three.js' USDZ convertor to be compatible with .png textures with transparency. By default, only .jpg textures without transparency is supported.

**Comment by elalish** [comment]
Awesome, thank you! And I bet three.js would be happy to have your fix for .png textures too.
